src/storage.rs

```diff
@@ -419,200 +450,324 @@ impl Cache {
+                // First try to get a cached copy from cargo's registry.
+                if let Some(reg) = self.cargo_registry.as_ref() {
+                    let fetched_src = reg.src().join(&dir_name);
+                    if fetch_is_ok(&fetched_src).await {
+                        return Ok(fetched_src);
+                    }
+                }
 
-        Ok(paths)
+                // Paths for the fetched package and checkout in our local cache.
+                let fetched_package = root
+                    .join(CACHE_REGISTRY_CACHE)
+                    .join(format!("{}.crate", dir_name));
+                let fetched_src = root.join(CACHE_REGISTRY_SRC).join(&dir_name);
+
+                // Check if the resource is already available in our local cache.
+                let fetched_package_ = fetched_package.clone();
+                let cached_file = tokio::task::spawn_blocking(move || {
+                    File::open(&fetched_package_).map(|file| {
+                        // Update the atime and mtime for this crate to ensure it isn't
+                        // collected by the gc.
+                        let now = filetime::FileTime::now();
+                        if let Err(err) =
+                            filetime::set_file_handle_times(&file, Some(now), Some(now))
+                        {
+                            warn!(
+                                "failed to update mtime for {}, gc may not function correctly: {}",
+                                fetched_package_.display(),
+                                err
+                            );
+                        }
+                        file
+                    })
+                })
+                .await?;
+
+                // If the file isn't in our local cache, make sure to download it.
+                let file = match cached_file {
+                    Ok(file) => file,
+                    Err(_) => {
+                        let network = network
+                            .as_ref()
+                            .expect("running as --frozen but needed fetches!");
+
+                        // We don't have it, so download it
+                        let url = Url::parse(&format!(
+                            "https://crates.io/api/v1/crates/{package}/{version}/download"
+                        ))?;
+                        info!(
+                            "downloading package {}:{} from {} to {}",
+                            package,
+                            version,
+                            url,
+                            fetched_package.display()
+                        );
+                        network.download_and_persist(url, &fetched_package).await?;
+
+                        // Check if the resource is already available in our local cache.
+                        let fetched_package_ = fetched_package.clone();
+                        tokio::task::spawn_blocking(move || File::open(&fetched_package_)).await??
```

Comment:
wait dafuq? We download and then check if already available? Is this comment just wrong?

Reply:
Yup, the comment is just wrong, I accidentally copied it.